src/h3lib/lib/algos.c

```diff
@@ -150,77 +150,97 @@ static const Direction NEW_ADJUSTMENT_III[7][7] = {
      CENTER_DIGIT, IJ_AXES_DIGIT}};
 
 /**
- * Maximum number of indices that result from the kRing algorithm with the given
+ * Maximum number of cells that result from the kRing algorithm with the given
  * k. Formula source and proof: https://oeis.org/A003215
  *
- * @param k k value, k >= 0.
+ * @param  k   k value, k >= 0.
  */
 int H3_EXPORT(maxKringSize)(int k) { return 3 * k * (k + 1) + 1; }
 
 /**
- * k-rings produces indices within k distance of the origin index.
+ * Produce cells within grid distance k the origin cell.
  *
- * k-ring 0 is defined as the origin index, k-ring 1 is defined as k-ring 0 and
- * all neighboring indices, and so on.
+ * k-ring 0 is defined as the origin cell, k-ring 1 is defined as k-ring 0 and
+ * all neighboring cells, and so on.
  *
  * Output is placed in the provided array in no particular order. Elements of
  * the output array may be left zero, as can happen when crossing a pentagon.
  *
- * @param origin Origin location.
- * @param k k >= 0
- * @param out Zero-filled array which must be of size maxKringSize(k).
+ * @param  origin   origin cell
+ * @param  k        k >= 0
+ * @param  out      zero-filled array which must be of size maxKringSize(k)
  */
 void H3_EXPORT(kRing)(H3Index origin, int k, H3Index* out) {
-    int maxIdx = H3_EXPORT(maxKringSize)(k);
-    int* distances = malloc(maxIdx * sizeof(int));
-    H3_EXPORT(kRingDistances)(origin, k, out, distances);
-    free(distances);
+    // Optimistically try the faster k-ring algorithm first
```

Comment:
Might be good to comment on the code duplication from `kRingDistances`.

Reply:
Actually, looking at it now, I think I can just put most of the logic into `kRingDistances`, making it handle a `NULL` value for `distances`.

Reply:
Updated!